Anisotropic crack propagation along cleavage planes in crystalline materials

adandeka opened the pull request:
Enhancement - Considers crack propagation along weak cleavage planes in a material
Issue # 14775
closes #14775

<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

adandeka committed:
Anisotropic crack propagation along cleavage planes in crystalline materials
Enhancement - Considers crack propagation along weak cleavage planes in a material
Issue # 14775
closes #14775

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/473826/) on 904cabd wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14824/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14824/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 5cf265aaaf4f12eb82709af329dc6423cd378a20`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/473849/) on 2e22497 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/14824/site)

This comment will be updated on new commits.

jiangwen84 commented:
@adandeka You can visualize your documentation using the link https://mooseframework.inl.gov/docs/PRs/14824/site/source/kernels/ACInterfaceBetaPenalty.html

Looks like some of your equations are not correctly shown there. Please do not use your owner symbol shortcut. The reference should list in bib file. You can follow ComputeLinearElasticPFFractureStress.md

jiangwen84 commented:
@adandeka Could you derive your ACInterfaceBetaPenalty from ACInterface to avoid code duplication?

adandeka commented:
@jiangwen84 
Yes, I think I can derive ACInterfaceBetaPenalty from ACInterface. Should I update the file and commit the change to this branch (feature_14775) ?
I will also look at the documentation file and update it.

jiangwen84 commented:
You can modify your existing PR and push to your origin, Github will automatically update your PR.

adandeka committed:
Removed previous ACInterfaceBetaPenalty files

adandeka committed:
ACInterfaceBetaPenalty files added
-- ACInterfaceBetaPenalty documentation file updated
-- ACInterfaceBetaPenalty kernel is derived from ACInterface
-- phase_field.bib file updated to include reference Clayton Knap 2015

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/474719/) on a4af488 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14824/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14824/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 5cf265aaaf4f12eb82709af329dc6423cd378a20`

adandeka commented:
@jiangwen84 I have updated all the files according to comments. Please let me know if I need to take any further action.

jiangwen84 commented:
@permcody Do you know what I can see this PR's documentation? 

The requested URL /docs/PRs/14824/site was not found on this server.

permcody commented:
We've been exceedingly tight on space lately. Having 30+ PRs open is not sustainable on our current hardware. Jason has had no choice but to delete the docs for open PRs to make enough space. Hopefully we can migrate to another server soon.

jiangwen84 commented on modules/phase_field/src/kernels/ACInterfaceBetaPenalty.C:
```diff
@@ -0,0 +1,85 @@
+Real
+ACInterfaceBetaPenalty::computeQpResidual()
+{
+  return (1 + _beta_penalty) * _grad_u[_qp] * kappaNablaLPsi() - betaNablaPsi();
```
@adandeka Could you give the detailed derivation for the weak form of the second term? Add them to the documentation.

jiangwen84 commented on modules/phase_field/src/kernels/ACInterfaceBetaPenalty.C:
```diff
@@ -0,0 +1,85 @@
+ACInterfaceBetaPenalty::betaNablaPsi()
+{
+  Real ngradc;
+  ngradc = _grad_u[_qp](0) * _cleavage_plane_normal[0] +
```
can be simplified as _grad_u[_qp] * _clevarge_plane_normal.

Same for the ngradpsi below.

adandeka replied:
@jiangwen84 
_grad_u[_qp] * _clevarge_plane_normal  did not work.
It gives me error:
'const value_type {aka const libMesh::VectorValue<double>}' is not derived from 'const std::complex<_Tp>'

The declaration in the header file is   const std::vector<Real> _cleavage_plane_normal;
Does this multiplication needs any other type ?

jiangwen84 commented on modules/phase_field/include/kernels/ACInterfaceBetaPenalty.h:
```diff
@@ -0,0 +1,33 @@
+class ACInterfaceBetaPenalty : public ACInterface
+{
+public:
+  ACInterfaceBetaPenalty(const InputParameters & parameters);
```
Maybe use a different name for this class. ACInterfaceCleavageFracture.

jiangwen84 commented on modules/phase_field/include/kernels/ACInterfaceBetaPenalty.h:
```diff
@@ -0,0 +1,33 @@
+  Real betaNablaPsi();
+
+  // penalty for damage on planes not normal to the weak (favoured) cleavage
+  // plane (Clayton & Knap, 2015)
```
Use Doxygen comment format. See  http://www.doxygen.nl/manual/docblocks.html

Single line can use ///

jiangwen84 commented on modules/phase_field/doc/content/source/kernels/ACInterfaceBetaPenalty.md:
```diff
@@ -0,0 +1,32 @@
+Its weak form is
+
+\begin{equation}
+\left( - \beta l_0 \left( \nabla^2 d - \nabla d \otimes \nabla d : \nabla \nabla d \right) \right) \psi,
```
Type here, \nabla d \otimes \nabla d  should be \nabla normal?

jiangwen84 reviewed (changes requested):
Good job. Looks good. I have a few moments.

jiangwen84 commented:
@adandeka Would you have time to work on this? Thanks.

adandeka commented:
@jiangwen84 
Thanks for checking. I am working on it. Will update the PR asap.

adandeka committed:
Removed files ACInterfaceBetaPenalty
Added files ACInterfaceCleavageFracture

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/494694/) on de9d245 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14824/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14824/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0b5c17986c68e79aab3610a5823e7c24e2d2bda3`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/494902/) on 82be3b8 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14824/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14824/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0b5c17986c68e79aab3610a5823e7c24e2d2bda3`

jiangwen84 commented on modules/phase_field/doc/content/source/kernels/ACInterfaceCleavageFracture.md:
```diff
@@ -0,0 +1,48 @@
+\int_{\Omega} \psi \left(\boldsymbol{n} \otimes \boldsymbol{n} : \nabla \nabla d \right) dV
+&= \int_{\Omega} \left( \nabla \cdot \left( \psi \boldsymbol{n} \left( \boldsymbol{n} \cdot \nabla d \right) \right) - \left( \boldsymbol{n} \cdot \nabla\psi \right) \left( \boldsymbol{n} \cdot \nabla d \right) - \left( \psi \nabla \cdot \boldsymbol{n} \right) \left( \boldsymbol{n} \cdot \nabla d \right) \right) dV
+\
+&= \int_{\Omega} \nabla \cdot \left( \psi \boldsymbol{n} \left( \boldsymbol{n} \cdot \nabla d \right) \right) dV - \int_{\Omega} \left( \boldsymbol{n} \cdot \nabla \psi \right) \left( \boldsymbol{n} \cdot \nabla d \right) dV - \int_{\Omega} \left( \psi \nabla \cdot \boldsymbol{n} \right) \left( \boldsymbol{n} \cdot \nabla d \right) dV
```
The notation \boldsymbol{n} is a bit confusing. Can you use M for cleavage normal and n for the boundary normal?

And you might want to replace `\` with `\\` to put equations on multiple lines.

adandeka replied:
Thanks for the comment. I changed the notation. Now it is, M for cleavage plane normal and n for boundary normal. Also, the multiline equation now uses \\.

jiangwen84 commented on modules/phase_field/src/kernels/ACInterfaceCleavageFracture.C:
```diff
@@ -0,0 +1,77 @@
+  }
+
+  Real ngradc;
+  ngradc = _grad_u[_qp] * _cleavage_plane_normal;
```
One line is good. 

Real ngradc = _grad_u[_qp] * _cleavage_plane_normal;

same for ngradphi

adandeka replied:
ngradphi is already written in terms of _cleavage_plane_normal
    ngradphi = _grad_phi[_j][_qp] * _cleavage_plane_normal;

Should it be written differently ?

jiangwen84 replied:
I mean you can combine two lines into one.

adandeka replied:
Thanks. I got it. Now, the two lines are combined.

jiangwen84 commented on modules/combined/test/tests/phase_field_fracture/crack2d_aniso_cleavage_plane.i:
```diff
@@ -0,0 +1,197 @@
+  [./ACInterfaceCleavageFracture]
+    type = ACInterfaceCleavageFracture
+    variable = c
+    beta_penalty = 1
```
minor issue, is beta_penalty = 1 sufficient? I thought it has to be very large.

adandeka replied:
beta=1 works well in cases where Gc ~ 2 J/m^2. If Gc is very high, then higher values of beta are required. The value in test input file is place holder.

jiangwen84 reviewed (changes requested):
Good job. Just minor changes.

adandeka committed:
ACInterfaceCleavageFracture.C updated

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/500086/) on 11080b9 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14824/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14824/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0b5c17986c68e79aab3610a5823e7c24e2d2bda3`

adandeka commented:
@jiangwen84 
Could you please help me understand this failed test (Mac Test) ? I did check the log but looked like some issue with libmesh.

adandeka committed:
ACInterfaceCleavageFracture updated with validParams

jiangwen84 commented:
@adandeka can you first rebase your branch? git --rebase upstream devel?

adandeka commented:
@jiangwen84 
Thanks for suggesting. I did rebase branch every time I updated files.
Now the Precheck fails in clang-format-check. I am applying clang-format package in atom but still it fails. Are the clang-format options for moose different than the standard package ?
Also, moosebuild used to generate a clang-format patch to fix any such issues but it does not generate it anymore.

jiangwen84 commented:
@permcody Do you know what causes precheck failure here? thanks.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/511076/) on 3d97879 : invalidated by @permcody

password expired

permcody commented:
Our bot's password expired. I just invalidated the target so it can run again.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/511076/) on 3d97879 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/14824/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/14824/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format f14a7fee63bd9a25d5f72f74f14a2513213a8516`

jiangwen84 commented:
@permcody could you active the remaining tests? Thanks.

jiangwen84 commented:
@permcody I do not have any further comments. If those tests pass, could you please merge them. Thanks.

adandeka commented:
@jiangwen84 
One quick question. Is there anything I need to do now ? Please let me know. Thanks.

jiangwen84 commented:
@adandeka No. Thanks for your great work. Sorry for taking so long. I look forward to your more contributions.

jiangwen84 commented:
@permcody could you merge it? Thanks.
